{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5849ba72",
   "metadata": {},
   "source": [
    "# Compute escape scores"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86dff6c5",
   "metadata": {},
   "source": [
    "## Import Python modules and initialize directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ca6fcca4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import glob\n",
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "import subprocess\n",
    "import multiprocessing\n",
    "\n",
    "resultsdir = '../results/predicted_escape_scores_v2/'\n",
    "if not os.path.isdir(resultsdir):\n",
    "    os.makedirs(resultsdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea25952f",
   "metadata": {},
   "source": [
    "## Compile input DMS data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62b1dbae",
   "metadata": {},
   "source": [
    "Read in processed data from `process_data.ipynb`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ea07007f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in data for 1603 antibodies\n"
     ]
    }
   ],
   "source": [
    "# Read in data on antibody sources\n",
    "escape_calc_resultsdir = '../results/processed_input_data/'\n",
    "ab_source_df = pd.read_csv(os.path.join(escape_calc_resultsdir, 'antibody_sources.csv'))\n",
    "abs_to_analyze = ab_source_df['antibody'].unique()\n",
    "print(f'Read in data for {len(ab_source_df)} antibodies')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31f68e6d",
   "metadata": {},
   "source": [
    "Read in the DMS data from Cao et al. Then, for each antibody from above, save a dataframe with the mutation-level escape scores for that antibody."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7ed86feb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>antibody</th>\n",
       "      <th>site</th>\n",
       "      <th>mutation</th>\n",
       "      <th>mut_escape</th>\n",
       "      <th>group</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>aa_mutation</th>\n",
       "      <th>max_mut_escape</th>\n",
       "      <th>norm_mut_escape</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BD-196</td>\n",
       "      <td>349</td>\n",
       "      <td>H</td>\n",
       "      <td>0.208866</td>\n",
       "      <td>C</td>\n",
       "      <td>S</td>\n",
       "      <td>S349H</td>\n",
       "      <td>0.507138</td>\n",
       "      <td>0.411853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BD-196</td>\n",
       "      <td>349</td>\n",
       "      <td>N</td>\n",
       "      <td>0.065818</td>\n",
       "      <td>C</td>\n",
       "      <td>S</td>\n",
       "      <td>S349N</td>\n",
       "      <td>0.507138</td>\n",
       "      <td>0.129783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BD-196</td>\n",
       "      <td>349</td>\n",
       "      <td>P</td>\n",
       "      <td>0.128169</td>\n",
       "      <td>C</td>\n",
       "      <td>S</td>\n",
       "      <td>S349P</td>\n",
       "      <td>0.507138</td>\n",
       "      <td>0.252731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BD-196</td>\n",
       "      <td>351</td>\n",
       "      <td>H</td>\n",
       "      <td>0.298263</td>\n",
       "      <td>C</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y351H</td>\n",
       "      <td>0.507138</td>\n",
       "      <td>0.588130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BD-196</td>\n",
       "      <td>352</td>\n",
       "      <td>G</td>\n",
       "      <td>0.071541</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A352G</td>\n",
       "      <td>0.507138</td>\n",
       "      <td>0.141069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191606</th>\n",
       "      <td>XGv-422</td>\n",
       "      <td>514</td>\n",
       "      <td>F</td>\n",
       "      <td>0.016126</td>\n",
       "      <td>E3</td>\n",
       "      <td>S</td>\n",
       "      <td>S514F</td>\n",
       "      <td>0.552973</td>\n",
       "      <td>0.029163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191607</th>\n",
       "      <td>XGv-422</td>\n",
       "      <td>514</td>\n",
       "      <td>I</td>\n",
       "      <td>0.023972</td>\n",
       "      <td>E3</td>\n",
       "      <td>S</td>\n",
       "      <td>S514I</td>\n",
       "      <td>0.552973</td>\n",
       "      <td>0.043351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191608</th>\n",
       "      <td>XGv-422</td>\n",
       "      <td>514</td>\n",
       "      <td>M</td>\n",
       "      <td>0.057837</td>\n",
       "      <td>E3</td>\n",
       "      <td>S</td>\n",
       "      <td>S514M</td>\n",
       "      <td>0.552973</td>\n",
       "      <td>0.104593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191609</th>\n",
       "      <td>XGv-422</td>\n",
       "      <td>514</td>\n",
       "      <td>R</td>\n",
       "      <td>0.020084</td>\n",
       "      <td>E3</td>\n",
       "      <td>S</td>\n",
       "      <td>S514R</td>\n",
       "      <td>0.552973</td>\n",
       "      <td>0.036321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191610</th>\n",
       "      <td>XGv-422</td>\n",
       "      <td>514</td>\n",
       "      <td>W</td>\n",
       "      <td>0.033401</td>\n",
       "      <td>E3</td>\n",
       "      <td>S</td>\n",
       "      <td>S514W</td>\n",
       "      <td>0.552973</td>\n",
       "      <td>0.060403</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>191611 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       antibody  site mutation  mut_escape group wildtype aa_mutation  \\\n",
       "0        BD-196   349        H    0.208866     C        S       S349H   \n",
       "1        BD-196   349        N    0.065818     C        S       S349N   \n",
       "2        BD-196   349        P    0.128169     C        S       S349P   \n",
       "3        BD-196   351        H    0.298263     C        Y       Y351H   \n",
       "4        BD-196   352        G    0.071541     C        A       A352G   \n",
       "...         ...   ...      ...         ...   ...      ...         ...   \n",
       "191606  XGv-422   514        F    0.016126    E3        S       S514F   \n",
       "191607  XGv-422   514        I    0.023972    E3        S       S514I   \n",
       "191608  XGv-422   514        M    0.057837    E3        S       S514M   \n",
       "191609  XGv-422   514        R    0.020084    E3        S       S514R   \n",
       "191610  XGv-422   514        W    0.033401    E3        S       S514W   \n",
       "\n",
       "        max_mut_escape  norm_mut_escape  \n",
       "0             0.507138         0.411853  \n",
       "1             0.507138         0.129783  \n",
       "2             0.507138         0.252731  \n",
       "3             0.507138         0.588130  \n",
       "4             0.507138         0.141069  \n",
       "...                ...              ...  \n",
       "191606        0.552973         0.029163  \n",
       "191607        0.552973         0.043351  \n",
       "191608        0.552973         0.104593  \n",
       "191609        0.552973         0.036321  \n",
       "191610        0.552973         0.060403  \n",
       "\n",
       "[191611 rows x 9 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in a complete list of sites and wildtype aminio acids for the RBD\n",
    "site_df = pd.read_csv('../data/RBD_sites.csv')\n",
    "site_df.rename(columns={\n",
    "    'site_SARS2' : 'site',\n",
    "    'amino_acid_SARS2' : 'wildtype'\n",
    "}, inplace=True)\n",
    "\n",
    "# Read in data and subset to filtered antibodies from above, then merge with\n",
    "# the dataframe giving the wildtype amino acid at each site\n",
    "dms_df = (\n",
    "    pd.read_csv('../convergent_RBD_evolution/use_res_clean.csv')\n",
    "    .query('antibody in @abs_to_analyze')\n",
    "    .merge(site_df[['site', 'wildtype']], how='left')\n",
    "    .assign(aa_mutation = lambda x: x['wildtype'] + x['site'].astype('str') + x['mutation'])\n",
    ")\n",
    "dms_df['max_mut_escape'] = dms_df.groupby('antibody')['mut_escape'].transform('max')\n",
    "dms_df['norm_mut_escape'] = dms_df['mut_escape'] / dms_df['max_mut_escape']\n",
    "\n",
    "# For each antibody, write a dataframe with amino-acid-level escape scores\n",
    "cao_dms_fs = []\n",
    "for (ab, data) in dms_df.groupby('antibody'):\n",
    "    output_f = os.path.join(resultsdir, f'{ab}_mut_effects.csv')\n",
    "    cao_dms_fs.append(output_f)\n",
    "    if not os.path.isfile(output_f):\n",
    "        data.to_csv(output_f, index=False)\n",
    "\n",
    "dms_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8533d2a9",
   "metadata": {},
   "source": [
    "## Compute escape scores from DMS data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d6c48f2",
   "metadata": {},
   "source": [
    "Make a list of input files from Bloom lab DMSs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c1c7839f",
   "metadata": {},
   "outputs": [],
   "source": [
    "bloom_dms_fs = [\n",
    "    '../ncov-dmsa/my_profiles/escape-frac-data/Dong_2021_AZD1061.csv',\n",
    "    '../ncov-dmsa/my_profiles/escape-frac-data/Dong_2021_AZD8895.csv',\n",
    "    '../ncov-dmsa/my_profiles/escape-frac-data/Starr_2021_LY-CoV555.csv',\n",
    "    '../ncov-dmsa/my_profiles/escape-frac-data/Starr_2021_LY-CoV016.csv',\n",
    "    '../ncov-dmsa/my_profiles/escape-frac-data/Starr_2021_REGN10933.csv',\n",
    "    '../ncov-dmsa/my_profiles/escape-frac-data/Starr_2021_REGN10987.csv'\n",
    " ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd31d733",
   "metadata": {},
   "source": [
    "For each antibody from both the Cao and Bloom datasets, use `dmsa-phenotype` to compute an escape score for each virus in an input multiple-sequence alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d7af8782",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define input params related to alignment\n",
    "alignment = '../ncov-dmsa/results/global-build/translations/aligned.gene.S_withInternalNodes.fasta'\n",
    "dms_wt_seq_id = 'Wuhan-Hu-1/2019'\n",
    "\n",
    "# Get DMS inputs\n",
    "cao_dms_inputs = [(f, 'norm_mut_escape', 'aa_mutation') for f in cao_dms_fs]\n",
    "bloom_dms_inputs = [(f, 'mut_escape', 'aa_mutation') for f in bloom_dms_fs]\n",
    "dms_inputs = cao_dms_inputs + bloom_dms_inputs\n",
    "\n",
    "# Cycle over each antibody with DMS data and compute escape scores\n",
    "commands = []\n",
    "for (i, (mut_effects_f, mut_effect_col, mutation_col)) in enumerate(dms_inputs, 1):\n",
    "\n",
    "    # Define output files\n",
    "    ab_label = os.path.basename(mut_effects_f).replace('_mut_effects.csv', '').replace('.csv', '')\n",
    "    output_json = os.path.join(resultsdir, f'{ab_label}_escape_pred.json')\n",
    "    output_df = os.path.join(resultsdir, f'{ab_label}_escape_pred.csv')\n",
    "    \n",
    "    # Run dmsa-pred\n",
    "    cmd = ' '.join([\n",
    "        'python',\n",
    "        '../ncov-dmsa/my_profiles/dmsa-pred/dmsa_pred.py phenotype-prediction',\n",
    "        '--model-type additive',\n",
    "        f'--alignment {alignment}',\n",
    "        f'--dms-wt-seq-id {dms_wt_seq_id}',\n",
    "        f'--mut-effects-df {mut_effects_f}',\n",
    "        f'--mut-effect-col {mut_effect_col}',\n",
    "        f'--mutation-col {mutation_col}',\n",
    "        '--mask-seqs-with-disallowed-aa-subs False',\n",
    "        f'--experiment-label {ab_label}',\n",
    "        f'--output-json {output_json}',\n",
    "        f'--output-df {output_df}'\n",
    "    ])\n",
    "    if not os.path.isfile(output_df):\n",
    "        commands.append(cmd)\n",
    "        #print(i, ab_label)\n",
    "        #! {cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "203cd2d2",
   "metadata": {},
   "source": [
    "Run the above list of commands in parallel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "943bfd3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_command(command):\n",
    "    \"\"\"Runs a given command in a subprocess.\"\"\"\n",
    "    process = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "    stdout, stderr = process.communicate()\n",
    "    return stdout.decode(), stderr.decode()\n",
    "\n",
    "with multiprocessing.Pool() as pool:\n",
    "    results = pool.map(run_command, commands)\n",
    "\n",
    "for stdout, stderr in results:\n",
    "    #print(\"Output:\", stdout)\n",
    "    if stderr:\n",
    "        print(\"Error:\", stderr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e6e1fe1",
   "metadata": {},
   "source": [
    "## Read in escape scores and merge with Nextstrain metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9c58e39e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in data for\n",
      "1609 antibodies\n",
      "3942 strains\n",
      "Saving data to ../results/predicted_escape_scores_v2/all_predictions.csv\n"
     ]
    }
   ],
   "source": [
    "# Read in predicted escape scores\n",
    "fs = glob.glob(os.path.join(resultsdir, '*_escape_pred.csv'))\n",
    "dfs = []\n",
    "for f in fs:\n",
    "    antibody = os.path.basename(f).replace('_escape_pred.csv', '')\n",
    "    df = pd.read_csv(f)\n",
    "    df = df[~df['strain'].str.contains('NODE')]\n",
    "    df['antibody'] = antibody\n",
    "    dfs.append(df)\n",
    "pred_df = pd.concat(dfs)\n",
    "pred_df.rename(columns={'pred_phenotype':'escape_score'}, inplace=True)\n",
    "\n",
    "# Read in metadata from Nextstrain\n",
    "metadata_f = '../ncov-dmsa/results/global-build/global-build_subsampled_metadata.tsv.xz'\n",
    "metadata_df = pd.read_csv(\n",
    "    metadata_f,\n",
    "    compression='xz', sep='\\t', on_bad_lines='skip'\n",
    ")\n",
    "metadata_df['date'] = pd.to_datetime(metadata_df['date'])\n",
    "metadata_df['year'] = metadata_df['date'].apply(\n",
    "    lambda x: x.year + ((x.dayofyear - 1) / 365)\n",
    ")\n",
    "metadata_df['time'] = metadata_df['year'] - metadata_df['year'].min()\n",
    "\n",
    "# Merge with prediction dataframe\n",
    "pred_df = pred_df.merge(\n",
    "    metadata_df[['strain', 'date', 'year', 'time', 'Nextstrain_clade']],\n",
    "    on='strain', how='left'\n",
    ")\n",
    "\n",
    "# Report stats and save data to file\n",
    "n_abs = len(pred_df['antibody'].unique())\n",
    "n_strains = len(pred_df['strain'].unique())\n",
    "assert sum(pred_df['strain'].value_counts() != n_abs) == 0\n",
    "print('Read in data for')\n",
    "print(n_abs, 'antibodies')\n",
    "print(n_strains, 'strains')\n",
    "\n",
    "output_f = os.path.join(resultsdir, 'all_predictions.csv')\n",
    "if not os.path.isfile(output_f):\n",
    "    print('Saving data to', output_f)\n",
    "    pred_df.to_csv(output_f, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8942ec34",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ncov-ab-escape",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
